### PR TITLE
Adding multiple role support in navigation

### PR DESCRIPTION
**Pull request by sharidas**
The navigation manager should support multiple
roles. It was allowed to enter only one role
in the info.xml. With this change, the app
developer can provide multiple roles comma
separted for role.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

It would be nice to have multiple roles. One scenario that came to my attention was with users page. The users page should be accessible for both admin and group admin. Hence this PR helps for the same.

## Related Issue

https://github.com/owncloud/core/issues/30922

## Motivation and Context

It would be nice to have multiple roles. One scenario that came to my attention was with users page. The users page should be accessible for both admin and group admin. Hence this PR helps for the same.

## How Has This Been Tested?

- [x] Verified that single roles and multiple roles with users page.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
Interesting. cc @pmaier1 


@sharidas what are roles ? is it an alias for groups ?

I don't remember seeing any concept of role in info.xml, would be great if you could clarify this. Maybe provide steps to demonstrate the behavior before and after your PR.

**Comment by sharidas** [comment]
When I tried using user_management app, I was trying to test if everything works as before. And to my surprise it was broken when user logged in as group admin. By broken I mean, the users page wont get displayed or cannot be accessed from the UI for group admin. Had a look at market app's info.xml which showed me https://github.com/owncloud/market/blob/master/appinfo/info.xml#L28. After debugging a bit figured out we could have single value for the role in info.xml. Significance of role is : to whom should the UI of the app should be displayed. So if the role is `admin` only user logged in as admin could see the UI. If the role is `sub-admin` then sub-admin user could see it.

This also brings a question what if an app's UI should be visible or usable for both admin and sub-admin. Hence extended the info.xml to use `role="admin,sub-admin"` ( ref: https://github.com/owncloud/user_management/pull/10/files#diff-224bb875c5d3cfcd7621135069013242R15). The change in the core would help app developers to provide comma separated role ( instead of single role ). Hope this make sense?

**Comment by PVince81** [comment]
Do you mean there is a regression where a group admin cannot use the users page any more ? If yes, please bisect to find when the regression was introduced and make a p1 ticket as this is quite important and could be a release blocker.

**Comment by DeepDiver1975** [comment]
> If yes, please bisect to find when the regression was introduced and make a p1 ticket as this is quite important and could be a release blocker.

no release blocker - this is master only

the sub.admin role was introduced by myself while moving the user management app out side of the core repo

the sub-admin role was never part of stable10 -> no release blocker

**Comment by sharidas** [comment]
> no release blocker - this is master only

Indeed this is not a release blocker as this issue should be there in master only.

**Comment by PVince81** [comment]
we still need a "regression" label and priority somewhere. would be good to have a ticket and clarify that it is master only

File: lib/private/NavigationManager.php [hunk 1]
```diff
@@ -145,13 +145,30 @@ private function init() {
...
+			//Remove whitespace in the role
+			$role = trim($role);
+			/**
+			 * Can have multiple roles like role="admin,sub-admin"
+			 */
+			$role = explode(',', $role);
+
+			$discontinue = true;
+			if ($this->isAdmin() && (in_array('admin', $role, true) === true)) {
+				$discontinue = false;
```
Comment (DeepDiver1975) [inline lib/private/NavigationManager.php]: just continue in here

File: lib/private/NavigationManager.php [hunk 2]
```diff
@@ -145,13 +145,30 @@ private function init() {
...
+
 			$role = isset($nav['@attributes']['role']) ? $nav['@attributes']['role'] : 'all';
-			if ($role === 'admin' && !$this->isAdmin()) {
-				continue;
+			//Remove whitespace in the role
+			$role = trim($role);
+			/**
+			 * Can have multiple roles like role="admin,sub-admin"
+			 */
+			$role = explode(',', $role);
```
Comment (DeepDiver1975) [inline lib/private/NavigationManager.php]: array elements need to be trimmed

File: tests/lib/NavigationManagerTest.php [hunk 3]
```diff
@@ -232,6 +236,22 @@ public function providesNavigationConfig() {
...
+				'active' => false
+			]], ['navigation' => ['@attributes' => ['role' => 'admin,sub-admin'], 'route' => 'test.page.index']], true],
+			'testSubadmin' => [[[
+				'id' => 'test',
+				'order' => 100,
+				'href' => '/apps/test/',
+				'icon' => '/apps/test/img/app.svg',
+				'name' => 'Test',
+				'active' => false
+			]], ['navigation' => ['@attributes' => ['role' => 'admin,sub-admin'], 'route' => 'test.page.index']], false, true]
```
Comment (DeepDiver1975) [inline tests/lib/NavigationManagerTest.php]: please add test for 'admin, sub-admin' and 'sub-admin' and 'admin'

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: lib/private/NavigationManager.php [hunk 4]
```diff
@@ -145,13 +145,30 @@ private function init() {
...
+			//Remove whitespace in the role elements
+			$role = array_map('trim', $role);
+
+			$shallContinue = false;
+			if ($this->isAdmin() && (in_array('admin', $role, true) === true)) {
+				$shallContinue = true;
```
Comment (sharidas) [inline lib/private/NavigationManager.php]: Code readability improved. ( no inverted logic ).

File: lib/private/NavigationManager.php [hunk 5]
```diff
@@ -145,13 +145,30 @@ private function init() {
...
+
 			$role = isset($nav['@attributes']['role']) ? $nav['@attributes']['role'] : 'all';
-			if ($role === 'admin' && !$this->isAdmin()) {
-				continue;
+			/**
+			 * Can have multiple roles like role="admin,sub-admin"
+			 */
+			$role = explode(',', $role);
+			//Remove whitespace in the role elements
+			$role = array_map('trim', $role);
```
Comment (sharidas) [inline lib/private/NavigationManager.php]: Trim the elements of array.

File: tests/lib/NavigationManagerTest.php [hunk 6]
```diff
@@ -232,6 +236,54 @@ public function providesNavigationConfig() {
...
+				'active' => false
+			]], ['navigation' => ['@attributes' => ['role' => 'admin'], 'route' => 'test.page.index']], true],
+			'testAdminWithSpace' => [[[
+				'id' => 'test',
+				'order' => 100,
+				'href' => '/apps/test/',
+				'icon' => '/apps/test/img/app.svg',
+				'name' => 'Test',
+				'active' => false
+			]], ['navigation' => ['@attributes' => ['role' => '   admin   '], 'route' => 'test.page.index']], true]
```
Comment (sharidas) [inline tests/lib/NavigationManagerTest.php]: Added tests for "admin,sub-admin", "admin" and "sub-admin". Added tests for roles with spaces.

**Review by DeepDiver1975 — APPROVED** [review]

**Pull request merged**